Feature(client): 대시보드 배너 구현 및 개발 워크플로우 개선

karnelll opened the pull request:
## 📌 Related Issues

> 관련된 Issue를 태그해주세요. (e.g. - close #25)

- close #80
- close #94

---

## ✅ 체크 리스트

- [x] PR 제목의 형식을 잘 작성했나요? e.g. Feat(client): 배너 컴포넌트 및 리소스 추가
- [x] 빌드가 성공했나요? (pnpm build)

---

## 📄 Tasks

- 대시보드 배너 컴포넌트 구현 (7단계)
- 배너 관련 SVG 및 리소스 추가
- 배너 이미지/도토리/치삐 단계별 그래픽 적용
- 관련 상수 및 유틸 분리
- 배너 컴포넌트 리팩토링 및 코드 품질 개선
- 중복된 이미지 파일명 로직 제거 및 BANNER_IMAGES 상수 활용
- 배열 인덱스 범위 검사 추가로 안전성 향상
- 매직 넘버 제거 및 위치 상수화로 유지보수성 개선
- Timer 컴포넌트 useEffect 의존성 배열 수정
- 메모 텍스트 3줄 초과 시 4번째 줄부터 완전히 잘리도록 CSS 강화 (`line-clamp` 직접 적용 및 overflow-hidden, max-height 등 추가)
- 데일리 리마인드 카드에 썸네일 없는 테스트 케이스(mock data) 추가
- 린트/포맷 자동수정 및 코드 스타일 일관성 유지
- 썸네일이 없을 때 배경 박스가 보이지 않도록 UI 개선

---

## ⭐ PR Point (To Reviewer)

- 배너 레이아웃 및 SVG 배치가 의도대로 잘 보이는지 확인 부탁드립니다.
- 도토리 개수에 따른 배너 상태 변화가 올바르게 작동하는지 확인해주세요.
- 배너 배경 이미지가 도토리 7개 달성 시에만 표시되는지 확인해주세요.
- 코드 리팩토링으로 인한 기능 변경이 없는지 확인해주세요.

---

## 🔧 주요 변경사항

### 새로운 기능
- 배너 컴포넌트 모듈화 (BannerHeadline 분리)
- 도토리 7개 달성 시 특별 배경 이미지 표시

### 코드 품질 개선
- Asset import 경로 통일 (@assets 별칭 사용)
- 중복 코드 제거 및 상수 활용
- 안전한 배열 인덱스 접근
- 매직 넘버 제거 및 위치 상수화

### 버그 수정
- FeedbackModal backdrop 클릭 처리
- Timer 컴포넌트 useEffect 의존성 배열 수정
- 치삐 이미지 로딩 문제 해결

---

## 🚀 개발 워크플로우 개선 안내

### 새로 추가된 스크립트
```bash
# 코드 포맷팅
pnpm format

# 린트 자동 수정
pnpm lint:fix

# 포맷팅 + 린트 자동 수정
pnpm fix-all

# 포맷팅 + 린트 자동 수정 + 빌드 (추천)
pnpm check-all
```

### 사용법
**커밋 전 필수 실행:**
```bash
pnpm check-all
```
이 명령어 하나로 포맷팅, 린트 자동 수정, 빌드가 순차적으로 실행됩니다.

### 개선된 점
- **자동화**: 수동으로 실행하던 작업들이 한 번에 처리됩니다.
- **일관성 유지**: 모든 개발자가 동일한 코드 스타일을 적용받습니다.
- **품질 보장**: 커밋 전에 코드 품질을 자동으로 검사합니다.
- **시간 절약**: 리뷰 시간 단축 및 실수 방지에 도움이 됩니다.

### 적용 대상
- client, extension, landing, design-system 전 앱 공통 적용
- Prettier + ESLint + Turbo 기준으로 안정적으로 실행됩니다

### ⚠️ 주의 사항
이번 작업 과정에서 `develop` 브랜치에 실수로 직접 푸시된 부분이 있었습니다. 해당 내용은 슬랙에 먼저 공유드렸고, 동일한 내용을 이번 PR에도 기록해두었습니다. 혼란을 드려 죄송합니다.

## 📷 Screenshot
<img width="1158" height="367" alt="스크린샷 2025-07-16 오전 3 13 46" src="https://github.com/user-attachments/assets/604359f6-0046-4e37-8942-db5d23f218e2" />
<img width="1224" height="363" alt="스크린샷 2025-07-16 오전 3 13 57" src="https://github.com/user-attachments/assets/79a5aaf5-2d62-44c9-afa8-49641eea2004" />
<img width="1223" height="365" alt="스크린샷 2025-07-16 오전 4 22 52" src="https://github.com/user-attachments/assets/d59f5919-19b9-4093-abe8-13e8dfff4cb3" />
<img width="1231" height="367" alt="스크린샷 2025-07-16 오전 3 14 24" src="https://github.com/user-attachments/assets/3572fbf3-73be-4cb8-b25d-05e5d16f12d7" />
<img width="1225" height="363" alt="스크린샷 2025-07-16 오전 3 14 38" src="https://github.com/user-attachments/assets/898a7163-f865-498c-9faa-b7a6a39df929" />
<img width="1233" height="364" alt="스크린샷 2025-07-16 오전 3 14 50" src="https://github.com/user-attachments/assets/7d6a1d9b-d881-485b-8a04-12a3b39ed859" />
<img width="1230" height="358" alt="스크린샷 2025-07-16 오전 3 15 01" src="https://github.com/user-attachments/assets/865c0a2c-e17b-41ed-8d41-299c6dcfc589" />
<img width="1226" height="361" alt="스크린샷 2025-07-16 오전 3 15 14" src="https://github.com/user-attachments/assets/af06f169-179a-4324-92d3-355ecf693596" />
<img width="1239" height="768" alt="스크린샷 2025-07-16 오전 3 15 44" src="https://github.com/user-attachments/assets/2b3db9f4-69b8-4ec6-8b63-aa043d660166" />

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 대시보드에 새로운 배너 컴포넌트와 BannerSection이 추가되어 도토리 개수에 따른 진행 상황을 시각적으로 확인할 수 있습니다.
  * 배너에 도토리 개수, 캐릭터 이미지, 레벨 정보, 타이머 등이 표시되며, 관련 하위 컴포넌트와 이미지 리소스도 포함됩니다.
  * 도토리 개수를 관리하고 갱신할 수 있는 훅과 API가 도입되었습니다.

* **개선 및 변경**
  * 대시보드 레이아웃이 개선되어 헤더가 고정되고 배너 및 섹션별 여백과 배치가 조정되었습니다.
  * 타이머 컴포넌트가 간소화되어 남은 시간만 표시하도록 변경되었습니다.
  * 여러 컴포넌트에서 에셋 경로가 alias(@assets) 기반으로 통일되었습니다.
  * 카드, 썸네일, 모달, 툴팁 등 UI 컴포넌트의 스타일과 동작이 일부 개선되었습니다.
  * 모달 닫기 동작이 외부 클릭으로도 가능하도록 변경되었습니다.
  * 배너 레이아웃 및 스타일링이 세분화되고, 모달 렌더링이 포탈 방식으로 변경되었습니다.

* **버그 수정 및 리팩터**
  * 텍스트 줄임 처리 방식이 개선되고, 조건부 렌더링 및 상태 관리가 최적화되었습니다.
  * 불필요한 주석과 미사용 코드가 정리되고, import 경로가 상대경로로 변경되어 코드 가독성이 향상되었습니다.
  * 일부 컴포넌트의 내부 로직과 상태 관리가 리팩터링되었습니다.

* **기타**
  * 타입 및 상수 정의가 추가되어 코드 일관성과 안정성이 향상되었습니다.
  * 목데이터가 보강되어 다양한 시나리오 테스트가 가능해졌습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat: 대시보드 배너 및 타이머 기능 임시 저장

Commit:
feat: 배너 컴포넌트 및 관련 리소스 추가

Commit:
refactor: 배너 컴포넌트 분리 및 모달 위치 개선

Commit:
refactor: 배너 컴포넌트 UI 개선 및 중앙 정렬 적용

Commit:
feat: 카드 컴포넌트 UI 개선 - 썸네일 크기 통일 및 제목 말줄임 처리

Commit:
refactor: 배너 컴포넌트 코드 품질 개선 및 기능 최적화

Commit:
refactor: 배너 컴포넌트 폴더 구조 개선 및 코드 품질 향상

Commit:
fix: 리마인드 카드 썸네일 배경색 추가로 중앙 정렬 개선

Commit:
refactor: 배너 관련 파일 폴더 구조 개선

Commit:
cleanup: 중복 파일 제거 및 폴더 구조 정리

Commit:
fix: 썸네일 없을 때 배경 박스 제거 및 메모 텍스트 3줄 제한 강화

Commit:
fix: 린트 오류 해결

Commit:
fix: BannerHeadline 컴포넌트 수정

jjangminii reviewed (commented):
2단계 텍스트 3-4단계랑 같은데 1단계랑 같게 바꿔주세요-!
<img width="249" height="400" alt="image" src="https://github.com/user-attachments/assets/77dc230b-a5ff-490d-80b0-ebeaf2a09e42" />

karnelll committed:
Merge branch 'develop' into feature/#80/banner-component

jjangminii reviewed (dismissed):
배너 분기처리 단계가 많아서 복잡했을텐데 고생하셨습니다-! 추후 상수로 관리해서 디자인 시스템 내 컨벤션으로 정의해두면 더 일관성 있게 사용할 수 있을 것 같아요-!!

Commit:
fix: 헤더 로고 직접 import 방식으로 변경 및 스타일 수정

constantly-dev commented on apps/client/src/shared/utils/styleUtils.ts:
```diff
@@ -34,7 +34,7 @@ export const CARD_CLASSES = {
   memo: 'w-[25.3rem] h-[9.6rem] pl-[4.7rem] pr-[6.7rem] py-[1.2rem] rounded-[0.6rem] flex items-center justify-center',
   memoInner:
-    'w-[22.9rem] h-[7.2rem] body2-m text-gray400 line-clamp-3 break-words self-start',
+    'w-[22.9rem] max-h-[7.2rem] body2-m text-gray400 break-words self-start overflow-hidden text-ellipsis [display:-webkit-box] [-webkit-line-clamp:3] [-webkit-box-orient:vertical] [line-height:1.6] [max-height:4.8rem]',
```
`max-height` 값 관련 값 체크하고 수정해주세요!

karnelll replied:
넵!!

constantly-dev commented on apps/client/src/shared/components/ui/modalPop/ModalPop.tsx:
```diff
@@ -7,7 +7,7 @@ import {
   ToggleButton,
 } from '@pinback/design-system/ui';
-import { POP_TEXTAREA_MAX_LENGTH } from '@constants/index';
+import { POP_TEXTAREA_MAX_LENGTH } from '@/constants/index';
```
`index` 파일은 경로에 포함하지 않아도 적용되지 않나요?!

karnelll replied:
빌드 오류 해결하다가 잘못 들어간 것 같아요!! 수정할게요!

constantly-dev commented on apps/client/src/shared/apis/timerApi.ts:
```diff
@@ -48,3 +49,8 @@ export const getBookmarkCards = async (
 };
+
+export const getAcornCount = async (): Promise<AcornCountResponse> => {
+  const response = await apiRequest.get<AcornCountResponse>('/api/acorn-count');
+  return response.data;
```
```suggestion
const { data } = await apiRequest.get<AcornCountResponse>('/api/acorn-count');
return data;
```

data만 return해서 사용할 것이라면 바로 받을 때 구조 분해 할당도 가능해요!

karnelll replied:
이번 API는 테스트 용도로 작성된 것이라, 실제 연동 시에는 응답 구조에 맞춰 구조 분해 할당 등 리팩토링하도록 하겠습니다!! PR에 미리 작성해두지 못해 죄송합니다 🙏

constantly-dev commented on apps/client/src/pages/dashboard/dashboard.tsx:
```diff
@@ -48,39 +48,34 @@ const Dashboard = () => {
   }, [mockBookmarkCards]);
-  const onPopUpOpen = () => {
-    setIsPopUpOpen(true);
+
+  const getBookmarkSectionMargin = () => {
+    if (hasBookmarks) {
+      return 'mt-[14rem]';
+    }
+    return 'mt-0';
```
해당 `mt-0`을 주는 것도 좋지만 `null`과 같이 아예 스타일을 주지 않는다는 의미로 작성하면 어떨까요?

karnelll replied:
좋은 것 같아요! 반영하겠습니다!

constantly-dev commented on apps/client/src/pages/dashboard/components/ui/indicators/Timer.tsx:
```diff
@@ -1,41 +1,41 @@
+const Timer = ({ onComplete: _onComplete, reminderHour = 9 }: TimerProps) => {
+  const [timeLeft, setTimeLeft] = useState<string>('00:00:00');
+
+  useEffect(() => {
+    const calculateTimeLeft = () => {
+      const now = new Date();
+      const reminderTime = new Date();
+      reminderTime.setHours(reminderHour, 0, 0, 0);
+
+      if (now.getHours() >= reminderHour) {
+        reminderTime.setDate(reminderTime.getDate() + 1);
+      }
+
+      const diff = reminderTime.getTime() - now.getTime();
+      const hours = Math.floor(diff / (1000 * 60 * 60));
+      const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));
+      const seconds = Math.floor((diff % (1000 * 60)) / 1000);
+
+      return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
+    };
+
+    setTimeLeft(calculateTimeLeft());
+
+    const interval = setInterval(() => {
+      setTimeLeft(calculateTimeLeft());
+    }, 1000);
+
+    return () => clearInterval(interval);
+  }, [reminderHour]);
+
+  return <span>{timeLeft}</span>;
```
해당 로직은 custom hook으로 빼도 좋을 것 같네요! 👍 그리고 diff나 hours 계산하는 로직을 하나로 묶어서 네이밍을 줘도 가독성이 좋아질 것 같습니다 굿!

karnelll replied:
좋은 피드백 감사합니다! 🙌
말씀 주신 대로 diff나 hours 계산 로직은 하나의 함수로 묶고, 의미 있는 네이밍을 부여해서 가독성을 높이도록 리팩토링해보겠습니다.
custom hook으로 분리하는 부분도 함께 고려해볼게요!

constantly-dev commented on apps/client/src/pages/dashboard/components/ui/indicators/Timer.tsx:
```diff
@@ -1,41 +1,41 @@
-
-  return <span className={`sub3-m ${getTimerColor()}`}>{getTimerText()}</span>;
+const Timer = ({ onComplete: _onComplete, reminderHour = 9 }: TimerProps) => {
+  const [timeLeft, setTimeLeft] = useState<string>('00:00:00');
```
`useState`에 초기 값을 넣어주면 `타입 추론`으로 인해서 따로 제니릭을 넣어주지 않아도 괜찮습니다! 물론 가독성을 위해서 넣는 것도 방법이지만, `타입 추론`을 적극적으로 사용하는 것도 좋은 방법이 될 것 같아요 👍 

```suggestion
const [timeLeft, setTimeLeft] = useState('00:00:00');
```

karnelll replied:
타입 추론을 활용하도록 수정해두었습니다! 감사합니다:)

constantly-dev commented on apps/client/src/pages/dashboard/components/ui/banner/BannerTimer.tsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+const BannerTimer = ({ acornCount }: BannerTimerProps) => {
+  if (acornCount !== 7) {
```
마지막 단계에 해당하는 상수 값을 만들어서 다른 곳에서도 재사용하면 좋을 것 같아요!

karnelll replied:
네!!

constantly-dev reviewed (commented):
분기 처리 많아서 힘드셨을텐데 수고하셨습니다~ 배너 완성되니까 진짜 마무리가 되는 느낌이네요!
코드 깔끔하게 정리해주신 것 같아요! 👍

Commit:
refactor: 코드 품질 개선 및 최적화

- MAX_ACORN_COUNT 상수 추가로 마지막 단계 기준값 재사용 가능
- Timer 로직을 useTimer custom hook으로 분리
- 타입 추론 활용으로 제네릭 타입 제거
- API 응답에서 구조 분해 할당 사용
- 충돌하는 max-height 스타일 수정
- 조건부 스타일링 개선

Commit:
fix: TimePicker 컴포넌트에서 불필요한 setErrorMessage prop 제거

Commit:
chore: prettier 및 린트 포맷팅 반영

Commit:
fix: pnpm-lock.yaml 충돌 해결 및 병합

jllee000 reviewed (approved):
고생하셨씁니당~~
